### PR TITLE
Beta support for Classroom World Map

### PR DESCRIPTION
**Pull request by basicer**
- Add map view and related tweaks to secret class options view.
- Refactor classroom settings to be schema driven
- Allow gems / level to be toggled.
- Stub campaign view static

### CODE REVIEW

File: app/templates/play/campaign-view.jade [hunk 1]
```diff
@@ -29,6 +29,9 @@ if view.showAds()
         em.spr powered by
         img(src="/images/pages/base/logo.png", title="Powered by CodeCombat - Learn how to code by playing a game ", alt="Powered by CodeCombat")
 
+    if view.shouldShow('back-to-classroom')
+      a.btn.btn-primary.under-logo(href='/play') Back to Classroom
```
Comment (sderickson) [inline app/templates/play/campaign-view.jade]: Needs translation

File: app/templates/play/campaign-view.jade [hunk 2]
```diff
@@ -220,26 +223,32 @@ if view.showAds()
...
+      if view.shouldShow('items')
+        button.btn.items(data-toggle='coco-modal', data-target='play/modal/PlayItemsModal', data-i18n="[title]play.items")
+      if view.shouldShow('heros')    
+        button.btn.heroes(data-toggle='coco-modal', data-target='play/modal/PlayHeroesModal', data-i18n="[title]play.heroes")
+      if view.shouldShow('achievements')
+        button.btn.achievements(data-toggle='coco-modal', data-target='play/modal/PlayAchievementsModal', data-i18n="[title]play.achievements")
+      if view.shouldShow('buy-gems')
         button.btn.gems(data-toggle='coco-modal', data-target='play/modal/BuyGemsModal', data-i18n="[title]play.buy_gems")
-      if !me.get('anonymous', true) && !features.codePlay
+      if !me.get('anonymous', true) && view.shouldShow('settings')
```
Comment (sderickson) [inline app/templates/play/campaign-view.jade]: Make sure this works as expected with a few feature modes. Good to refactor, but want to make sure third parties see no difference in behavior.

File: app/views/play/CampaignView.coffee [hunk 3]
```diff
@@ -123,6 +127,57 @@ module.exports = class CampaignView extends RootView
...
+            console.log "Sync", existing
+            for k,v of @courseLevels.toArray()
+              idx = v.get('original')
+              @courseLevelsFake[idx] = existing[idx]
+              @courseLevelsFake[idx].courseIdx = parseInt(k)
+              @courseLevelsFake[idx].requiresSubscription = false
+            console.log "Fake is ", @courseLevelsFake
+
+      )
+
```
Comment (sderickson) [inline app/views/play/CampaignView.coffee]: Logs to remove throughout this

File: app/views/play/CampaignView.coffee [hunk 4]
```diff
@@ -408,21 +467,26 @@ module.exports = class CampaignView extends RootView
       return me.getCampaignAdsGroup() is 'leaderboard-ads'
     false
 
-  annotateLevels: (orderedLevels) ->
+  annotateLevels: (orderedLevels) =>
```
Comment (sderickson) [inline app/views/play/CampaignView.coffee]: Why fat arrow?

File: app/views/play/CampaignView.coffee [hunk 5]
```diff
@@ -480,7 +547,12 @@ module.exports = class CampaignView extends RootView
     leaderboardModal = new LeaderboardModal supermodel: @supermodel, levelSlug: levelSlug
     @openModalView leaderboardModal
 
-  determineNextLevel: (orderedLevels) ->
+  determineNextLevel: (orderedLevels) =>
+    if @courseStats?
+      @applyCourseLogicToLevels(orderedLevels)
+      return true
```
Comment (sderickson) [inline app/views/play/CampaignView.coffee]: Also why fat arrow

File: app/views/play/CampaignView.coffee [hunk 6]
```diff
@@ -1008,3 +1081,83 @@ module.exports = class CampaignView extends RootView
...
+    nextSlug = @courseStats.levels.next.get('slug')
+    courseOrder = _.sortBy orderedLevels, 'courseIdx'
+    found = false
+    for level, levelIndex in courseOrder
+      playerState = @levelStatusMap[level.slug]
+      level.color = 'rgb(255, 80, 60)'
+      level.disabled = false
+
+      if level.slug is nextSlug
+        console.log "Next for you is #{level.name}", level
```
Comment (sderickson) [inline app/views/play/CampaignView.coffee]: console.log

File: app/views/play/CampaignView.coffee [hunk 7]
```diff
@@ -1008,3 +1081,83 @@ module.exports = class CampaignView extends RootView
 
   onClickBrainPopReplayButton: ->
     api.users.resetProgress().then(=> document.location.reload())
+
+  getLevels: () ->
+    return @courseLevelsFake if @courseLevels?
```
Comment (sderickson) [inline app/views/play/CampaignView.coffee]: I'm not sure what this name means, `courseLevelsFake`. What's its purpose exactly?

File: app/templates/courses/classroom-settings-modal.jade [hunk 8]
```diff
@@ -59,6 +59,18 @@ block modal-body-content
...
+      .form-group
+        label Student Experience Options
+        - var settings = view.schema.properties.settings.properties
+        for opt,name in settings
+          if name != "optionsEditable" || me.isAdmin()
+          .form-group
+            div.checkbox
+              input(type="checkbox", id="settings_"+name, name="settings/"+name, checked=view.classroom.getSetting(name))
+              label(for="settings_" + name)
+                span= opt.description
```
Comment (basicer) [inline app/templates/courses/classroom-settings-modal.jade]: These settings are just for us.  If we ship a more general teacher options system we'll have to translate it.

File: app/templates/courses/courses-view.jade [hunk 9]
```diff
@@ -116,25 +121,36 @@ mixin course-instance-body(courseInstance, classroom)
...
-        span(data-i18n="common.continue")
+      - var mapurl = view.urls.courseWorldMap({course: course, courseInstance: courseInstance});
+      if classroom.getSetting('map')
+        a.play-btn.btn.btn-forest.btn-lg.m-b-1(href=mapurl, data-level-slug=next.get('slug'), data-event-action="Students Continue Course")
+          span(data-i18n="common.continue")
+      else
+        - var url = view.urls.courseLevel({level: view.originalLevelMap[next.get('original')] || next, courseInstance: courseInstance});
+        a.play-btn.btn.btn-forest.btn-lg.m-b-1(href=url, data-level-slug=next.get('slug'), data-event-action="Students Continue Course")
+         span(data-i18n="common.continue")
+
```
Comment (basicer) [inline app/templates/courses/courses-view.jade]: I plan on doing some redecorating, so seemed best just to split them out.

File: app/models/Classroom.coffee [hunk 10]
```diff
@@ -194,3 +194,12 @@ module.exports = class Classroom extends CocoModel
...
     @fetch(options)
+
+  getSetting: (name) =>
+    settings = @get('settings') or {}
+    propInfo = Classroom.schema.properties.settings.properties
+    return settings[name] if name in Object.keys(settings)
+    if name in Object.keys(propInfo)
+      return propInfo[name].default
+
+    return false
```
Comment (basicer) [inline app/models/Classroom.coffee]: Tell me about it :(

File: app/views/play/CampaignView.coffee (hunk 7 above)
Comment (basicer) [inline app/views/play/CampaignView.coffee]: This helps us apply the existing nextLevel logic from the classroom on top of the world map.

**Pull request merged**